test/test_sparse.py

```diff
@@ -2678,6 +2679,103 @@ def sparse_log(x):
         test_op(3, 100, [3, 4, 2, 3, 5, 2])
         test_op(4, 100, [3, 4, 2, 3, 5, 2])
 
+    @cpu_only
+    def test_sparse_matmul(self):
+
+        def _indices2csr(indices, dim):
+            nnz = len(indices)
+            r = [0] * (dim + 1)
+            last_i = 0
+            for i in indices:
+                if i == last_i:
+                    r[last_i + 1] += 1
+                else:
+                    for _i in range(last_i, i + 1):
+                        r[_i + 1] = r[last_i + 1]
+                    last_i = i
+                    r[last_i + 1] += 1
+            for _i in range(last_i, dim):
+                r[_i + 1] = r[last_i + 1]
+            assert r[-1] == nnz
+            return r
+
+        def sparse_mm(a, b):
+            if a.shape[1] != b.shape[0]:
+                print("Incompatible matrices")
+                return None
+            n, p = a.shape
+            p, m = b.shape
+            indices_a = a._indices()
+            values_a = a._values()
+            indices_b = b._indices()
+            values_b = b._values()
+            nnz1 = len(indices_a[0])
+            nnz2 = len(indices_b[0])
+
+            if a.is_coalesced() and b.is_coalesced():
+                r2 = _indices2csr(indices_b[0], b.shape[0])
+                d = defaultdict(values_b.numpy().dtype.type)
+                for n1 in range(nnz1):
+                    for n2 in range(r2[indices_a[1][n1]], r2[indices_a[1][n1] + 1]):
+                        d[indices_a[0][n1].item(), indices_b[1][n2].item()] += values_a[n1] * values_b[n2]
+
+            else:
+                d = defaultdict(values_b.numpy().dtype.type)
+                for n1 in range(nnz1):
+                    for n2 in range(nnz2):
+                        if indices_b[0][n2] == indices_a[1][n1]:
+                            d[indices_a[0][n1].item(), indices_b[1][n2].item()] += values_a[n1] * values_b[n2]
+
+            i3 = []
+            j3 = []
+            values = []
+            for i, j in sorted(d):
+                i3.append(i)
+                j3.append(j)
+                values.append(d[i, j])
+            return torch.sparse_coo_tensor(torch.tensor([i3, j3]), torch.tensor(values), (n, m))
+
+        def test_op(sparse_dims, nnz, shape_a, shape_b):
+            a, i_a, v_a = self._gen_sparse(sparse_dims, nnz, shape_a)
+            b, i_b, v_b = self._gen_sparse(sparse_dims, nnz, shape_b)
+            a_dense = a.to_dense()
+            b_dense = b.to_dense()
+            r1 = a_dense.matmul(b_dense)
+
+            # python implementation
+            r2 = sparse_mm(a, b)
+            self.assertEqual(r1, r2.to_dense())
+            self.assertEqual((r2._values() != 0).all(), True)
+
+            # cpp implementation
+            r2 = torch.sparse.mm(a, b)
+            self.assertEqual(r1, r2.to_dense())
+
+            # check autograd support on sparse matmul
```

Comment:
This is probably a good place to start a new test ("test_sparse_matmul_grad"?) to improve test granularity

Reply:
Also, Future Work: implement gradcheck (and gradgradcheck) support for sparse ops